Fix bc scale with k endpoints

eriknw opened the pull request:
Fix `betweenness_centrality` rescaling given `k` and `endpoints=False`.

  Previously, normalizing could give values above 1:
  ```python
  >>> nx.betweenness_centrality(nx.star_graph(4), k=1, seed=1, endpoints=False)
  {0: 1.25, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}
  ```

Odd behavior with rescaling was reported in cugraph by @zmahoor in https://github.com/rapidsai/cugraph/issues/4941

Credit to @ChuckHastings for understanding the issue and identifying correct behavior. Are there any other tests you think should be added?

Note that `betweenness_centrality_subset` and `bipartite.betweenness_centrality` don't have `k`, so they don't have this issue. I haven't considered whether `edge_betweenness_centrality` could have a similar issue.

CC @rlratzel

eriknw commented on networkx/algorithms/centrality/tests/test_betweenness_centrality.py:
```diff
@@ -288,6 +288,17 @@ def test_directed_path_normalized(self):
 
+    def test_scale_with_k_on_star_graph(self):
+        G = nx.star_graph(4)
+        b = nx.betweenness_centrality(G, k=1, seed=1, endpoints=False)
+        assert b == pytest.approx({0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0})
```
Previous result:
```python
{0: 1.25, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}
```

eriknw commented on networkx/algorithms/centrality/tests/test_betweenness_centrality.py:
```diff
@@ -288,6 +288,17 @@ def test_directed_path_normalized(self):
+        assert b == pytest.approx({0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0})
+        b = nx.betweenness_centrality(G, k=1, seed=1, endpoints=True)
+        assert b == pytest.approx({0: 1.0, 1: 1.0, 2: 0.25, 3: 0.25, 4: 0.25})
+        b = nx.betweenness_centrality(G, k=1, seed=1, endpoints=False, normalized=False)
+        assert b == pytest.approx({0: 6.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0})
```
Previous result:
```python
{0: 7.5, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}
```
Somebody should double-check that the new results are correct! They _seem_ right to me.

ChuckHastings commented:
> Note that `betweenness_centrality_subset` and `bipartite.betweenness_centrality` don't have `k`, so they don't have this issue. I haven't considered whether `edge_betweenness_centrality` could have a similar issue.

`edge_betweenness_centrality` has no `endpoints` parameter, so the scaling logic is straightforward.

eriknw commented:
I found more worms in this can!

Question for everybody: is BC given `k` w/o normalization supposed to approximate the _full_ BC or not?

Previously, when given `k` and `normalized=False`, the BC results would be rescaled using `k` _only_ for undirected graphs (thereby giving an estimate of the full BC), but not for directed graphs (thereby computing BC for `k` samples), so one of these cases must be wrong.

So, assuming we want to compute or estimate the full BC, I expanded out to give the explicit scale factors for 12 of the 16 cases from the cartesian product of:
- normalized : {True, False}
- endpoints : {True, False}
- k : {None, k}
- directed : {True, False}

The missing 4 cases are b/c the scale factors are the same for `normalized=True` and `directed={True|False}`.

@ChuckHastings I know you are working on this as well in libcugraph. What do you think? Do our scale factors agree? What is generally desired when computing BC with `k` (full BC or BC for `k` samples; both seem reasonable to me)?

eriknw commented on networkx/algorithms/centrality/tests/test_betweenness_centrality.py:
```diff
@@ -288,6 +288,40 @@ def test_directed_path_normalized(self):
             assert b[n] == pytest.approx(b_answer[n], abs=1e-7)
 
+    @pytest.mark.parametrize(
+        ("normalized", "endpoints", "is_directed", "k", "expected"),
+        [
+            (True, True, True, None, {0: 1.0, 1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4}),
+            (True, True, True, 1, {0: 1.0, 1: 1.0, 2: 0.25, 3: 0.25, 4: 0.25}),
+            (True, True, False, None, {0: 1.0, 1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4}),
+            (True, True, False, 1, {0: 1.0, 1: 1.0, 2: 0.25, 3: 0.25, 4: 0.25}),
+            (True, False, True, None, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (True, False, True, 1, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (True, False, False, None, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (True, False, False, 1, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (False, True, True, None, {0: 20.0, 1: 8.0, 2: 8.0, 3: 8.0, 4: 8.0}),
+            (False, True, True, 1, {0: 20.0, 1: 20.0, 2: 5.0, 3: 5.0, 4: 5.0}),
+            (False, True, False, None, {0: 10.0, 1: 4.0, 2: 4.0, 3: 4.0, 4: 4.0}),
+            (False, True, False, 1, {0: 10.0, 1: 10.0, 2: 2.5, 3: 2.5, 4: 2.5}),
+            (False, False, True, None, {0: 12.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (False, False, True, 1, {0: 12.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (False, False, False, None, {0: 6.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+            (False, False, False, 1, {0: 6.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
+        ],
```
**DO NOT MERGE**

The diff in this comment shows the results from the main branch. Results only change when integer `k` is used:
```suggestion
        ("normalized", "endpoints", "is_directed", "k", "expected"),
        [
            (True, True, True, None, {0: 1.0, 1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4}),
            (True, True, True, 1, {0: 1.0, 1: 1.0, 2: 0.25, 3: 0.25, 4: 0.25}),
            (True, True, False, None, {0: 1.0, 1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4}),
            (True, True, False, 1, {0: 1.0, 1: 1.0, 2: 0.25, 3: 0.25, 4: 0.25}),
            (True, False, True, None, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (True, False, True, 1, {0: 1.25, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (True, False, False, None, {0: 1.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (True, False, False, 1, {0: 1.25, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (False, True, True, None, {0: 20.0, 1: 8.0, 2: 8.0, 3: 8.0, 4: 8.0}),
            (False, True, True, 1, {0: 4.0, 1: 4.0, 2: 1.0, 3: 1.0, 4: 1.0}),
            (False, True, False, None, {0: 10.0, 1: 4.0, 2: 4.0, 3: 4.0, 4: 4.0}),
            (False, True, False, 1, {0: 10.0, 1: 10.0, 2: 2.5, 3: 2.5, 4: 2.5}),
            (False, False, True, None, {0: 12.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (False, False, True, 1, {0: 3.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (False, False, False, None, {0: 6.0, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
            (False, False, False, 1, {0: 7.5, 1: 0.0, 2: 0.0, 3: 0.0, 4: 0.0}),
        ],
```

ChuckHastings commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -357,25 +359,47 @@ def _accumulate_edges(betweenness, S, P, sigma, s):
+            if directed:
+                scale = None
+            else:
+                scale = 0.5
```
I believe the original logic... and this logic... is incorrect.  The factor of 2 (or 0.5) is wrong.

From the Wikipedia article: `The division is done by (N−1)(N−2) for directed graphs (N-1)(N-2)/2 undirected graphs`.  This is predicated on an assumption that when you process an undirected graph that you iterate over all of the unique vertex pairs (when computing exact BC).  This can be done on an undirected graph because the shortest path from `(u,v)` is exactly reversed from the shortest path from `(v,u)`.  On a directed graph that's not guaranteed.  So the assumption (in computing this normalization factor) is that you've only counted the vertices on the path `(u,v)` and not counted the vertices on the path `(v,u)`.

In the nx implementation of exact betweenness, there's no logic to skip computing the path `(u,v)` and the path `(v,u)`.  In computing approximate betweenness we don't make that distinction either.

Consequently, the factor of 2 (or 0.5) should not be included.

eriknw replied:
Thanks @ChuckHastings. However, your explanation led me to the opposite conclusion. As you said, for undirected graphs, shortest paths are determined and counted for _both_ (u, v) and (v, u) pairs. Only one should be counted though, which is why dividing by two here is correct. This also explains why the normalization factor is the same for both directed and undirected graphs when `normalized=True`.

ChuckHastings replied:
Let me try to explain using an example.  Consider your test case here.  Let's take the specific case of False, False, True in your tests above.  So we're not normalizing, we're not including endpoints.

In an exact BC run, we're going to run BFS starting from each of the 5 vertices.
* From vertex 0, all paths are of length 1 and every vertex is an endpoint... so nothing gets added to the betweenness score
* From the other 4 vertices, there are 3 paths are of length 2, traveling through vertex 0 and arriving at each of the other vertices in the star.  Here's the paths: [(1, 0, 2), (1, 0, 3), (1, 0, 4), (2, 0, 1), (2, 0, 3), (2, 0, 4), (3, 0, 1), (3, 0, 2), (3, 0, 4), (4, 0, 1), (4, 0, 2), (4, 0, 3)].  Thus we end up with a value of 12 in the raw BC score for vertex 0 and a raw BC score for each of the other vertices.

Now imagine an implementation that operated on an undirected graph.  To reduce the computation, you don't compute all of the shortest paths, you only compute half of the paths... because you don't need to compute them all... it's symmetric.  Here would be the paths: [(1, 0, 2), (1, 0, 3), (1, 0, 4), (2, 0, 3), (2, 0, 4), (3, 0, 4)].  Thus we end up with a value of 6 in the raw BC score for vertex 0...  which is why we need to double that score...

Ultimately, the intention of that doubling factor is to make the results from operating on an undirected graph the same as if you were operating on the directed graph equivalent for that graph.  You want the results from the undirected graph [(0, 1), (0, 2), (0, 3), (0,4)] to be exactly the same as the results from the directed graph [(0, 1), (1, 0), (0, 2), (2, 0), (0, 3), (3, 0), (0, 4), (4, 0)] because they are different representations of exactly the same graph structure.


But networkx (and cugraph) are computing the raw BC score as if the graph were directed - even if the graph is undirected.  So we already get the raw score of 12 even if the graph is undirected, there's no reason to modify it.

ChuckHastings replied:
Or... is the goal to compute it using the directed graph method as I've described above and then correct the raw score (if we're not normalizing) to pretend that we computed it using the undirected graph optimization.

If that's the goal than we need the factor of 2 logic... although I'd have to review the whole flow again.

eriknw replied:
Based on my understanding, I believe your last comment is "correct". When computing non-normalized BC for undirected graphs without endpoints, there are the six paths that you identified. There is no need to double 6 to 12 to match non-normalized BC for directed graphs with the same bidirectional edges. The graph structure--for the example, the adjacency array--may be the same for directed or undirected, but the interpretation of "valid pairs" are different. For directed, `(s, t)` and `(t, s)` are considered; for undirected, only the set `{s, t}` makes a valid _unordered_ pair (so `(s, t)` and `(t, s)` are the same and should be counted only once).

eriknw replied:
To be clear (and repeat what we said), the implementation in networkx computes all 12 paths you identified in this case for the undirected graph because it counts both `(s, t)` and `(t, s)`, so it divides the non-normalized BC by 2 in order to count `{s, t}` once.

ChuckHastings commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -357,25 +359,47 @@ def _accumulate_edges(betweenness, S, P, sigma, s):
+            if directed:
+                scale = n / k
+            else:
+                scale = n / (k * 2)
```
same concern here.

ChuckHastings commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -357,25 +359,47 @@ def _accumulate_edges(betweenness, S, P, sigma, s):
+            if directed:
+                scale = None
+            else:
+                scale = 0.5
```
same concern here.

ChuckHastings commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -357,25 +359,47 @@ def _accumulate_edges(betweenness, S, P, sigma, s):
+            if directed:
+                scale = (n - 1) / k
+            else:
+                scale = (n - 1) / (2 * k)
```
Same concern here.

ChuckHastings replied:
Also... the test results below should be modified to reflect this.

ChuckHastings reviewed (approved):
I think we're on the same page now.

eriknw commented:
@dschult here is the minimal diff from main that passes the tests in this PR:
```diff
diff --git a/networkx/algorithms/centrality/betweenness.py b/networkx/algorithms/centrality/betweenness.py
index 42e09771d..c90ccd910 100644
--- a/networkx/algorithms/centrality/betweenness.py
+++ b/networkx/algorithms/centrality/betweenness.py
@@ -127,6 +127,8 @@ def betweenness_centrality(
        https://doi.org/10.2307/3033543
     """
     betweenness = dict.fromkeys(G, 0.0)  # b[v]=0 for v in G
+    if k == len(G):
+        k = None
     if k is None:
         nodes = G
     else:
@@ -372,12 +374,16 @@ def _rescale(betweenness, n, normalized, directed=False, k=None, endpoints=False
         if not directed:
             scale = 0.5
         else:
-            scale = None
+            scale = 1
     if scale is not None:
         if k is not None:
-            scale = scale * n / k
-        for v in betweenness:
-            betweenness[v] *= scale
+            if endpoints:
+                scale *= n / k
+            else:
+                scale *= (n - 1) / k
+        if scale != 1:
+            for v in betweenness:
+                betweenness[v] *= scale
     return betweenness


```
I explicitly expanded out the different cases to make it easier to see and consider each case. This helped me and @ChuckHastings communicate and understand each case (@ChuckHastings is updating cugraph in https://github.com/rapidsai/cugraph/pull/4974, and we're now in agreement--thanks @ChuckHastings!).

I just added some code comments to explain rescaling, because it's actually kind of nuanced, but straightforward if given careful thought of the definition of BC. If you prefer the minimal diff just above, I can update the PR. Computing `scale` in two passes with some state ("why do we set `scale = 1`?!") seems more difficult to understand, because a reader will need to unravel the procedural logic to understand what is happening for a particular case. Having a single pass if-else is a little longer, but clearer IMHO.

Summary of fixes:
- When `k` is not None and `endpoints` is False, scale by `(n - 1)` instead of `n`
- When `k` is not None and `normalized` is False and directed graph
  - scale by `n / k` if `endpoints` is True
  - scale by `(n - 1) / k` if `endpoints` is False
- Fix edge case when `k=n` and `endpoints` is False

eriknw commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -357,25 +362,61 @@ def _accumulate_edges(betweenness, S, P, sigma, s):
-            scale = 0.5
+            # Exclude endpoints and count the number of (s, t) node pairs
+            # that could have a path pass through v where v is not s or t.
+            if k is None:
+                scale = 1 / ((n - 1) * (n - 2))
+            else:
+                scale = 1 / (k * (n - 2))
```
> Fix edge case when k=n and endpoints is False

Here is where this fix comes into play. If `k = n`, these scales would be different for each of these branches. If `k = n`, the `k is None` branch should be used, because the `else` branch violates the assumption that `v` is not `s` or `t`.

dschult commented:
In terms of simplifying the code: we could take advantage of the various places we replace n with n-1. That reduces the number of checks for `endpoints`.  Something like:
```python
def _rescale(betweenness, n, normalized, directed=False, k=None, endpoints=False):
    N = n if endpoints else n - 1  # could put this in the call arg 'n' to _rescale and remove endpoints kwarg 
    if N <= 1:
        return betweenness
    if normalized:
        scale = 1 / (N * (N - 1))
    else:
        scale = 1 if directed else 0.5

    if k is not None and k < N:
        scale *= N / k
    if scale != 1:
        for v in betweenness:
            betweenness[v] *= scale
    return betweenness
```

Also, we might consider `k==n` as a mistake and raise an exception indicating that this isn't a sample -- it is the whole population. Making the special case `k==n` raise an exception might(?) be better for reducing mistakes by users.  But that can be a separate PR for sure.

I'm shooting for some code that describes the following ideas rather than going through every combination of vars. The ideas feel orthogonal -- but that may be just me:
- `endpoints` reduces the effective size of scaling from using n to n-1.
- `normalized` divides by n*(n-1) -- i.e. the number of possible pairs.
- unnormalized undirected counts need to be divided by 2
- scale up the sample to full BC using factor n/k

eriknw commented:
Certainly @dschult, and it's not just you. I find the expanded cases were useful for development and careful review, but I'm also inclined to simplify and condense the logic before merging. The previous two commits take steps to condense the logic similar to your suggestions; you may look at them both.

dschult reviewed (commented):
Nice -- thanks for breaking this out into comment-able chunks. And commenting them! :}

The case with k < n and endpoints is False still confuses me. It seems like `the number of valid (s, t) node pairs that could have a path through v where s != t` should depend on whether node `v` was in the sample or not in the sample. The largest possible count for nodes in the sample (when endpoints is False) is (k-1)*(n-2), but for nodes not in the sample the count can be `k*(n-2)`. Why are we using `k*(n-2)` for all the nodes?

Also, I still don't like the final `scale` computation because it requires learning 4 new values no matter what your particular case is. What do you think of adjusting `scale` after each comment paragraph under the conditions stated in that paragraph. For example after the undirected correct is discussed we could end with:
```python
    # should be counted only once. Dividing by 2 corrects for this.
    if not normalized and not directed:
        scale /= 2
```
Then to figure out `scale` for any given case we check the if-conditions to see the various impacts of different paragraphs. I'm hoping that is simpler than trying to hold and construct 4 values in my head. But that could just be me...

I also have one wording tweak: may -> must
Thanks!

dschult commented:
> Yeah, this is a great question. I need to think on it more carefully. The original code didn't do this either, so I usually try to minimize the changes done in a PR.

Agreed!  That's a bigger question and should be in another PR.  I'll raise a separate issue.

The code looks quite readable now (at least to me)!  Thank you!!

I believe the only question remaining is whether to allow `k>n` by setting `k = None` in that case, or to only set `k = None` for the `k==n` case and have `sample` raise for `k>n` case. The previous code raised for `k > n`. I have a slight preference for raising when the sample is larger than the population size. I'm fine going ahead and computing the population value when the `k=n`. That is what the previous code did.

Sorry for making this take so long -- I have learned a bunch, but slowed this down for sure...

eriknw commented:
> Sorry for making this take so long -- I have learned a bunch, but slowed this down for sure...

Nonsense! The code was begging for a "simplicity shakedown", so thanks for being on the journey for me :D  . And what better time to learn about these changes than when it's fresh in my mind? I've learned a lot too.

> I believe the only question remaining is whether to allow k>n by setting k = None in that case, or to only set k = None for the k==n case and have sample raise for k>n case. The previous code raised for k > n. I have a slight preference for raising when the sample is larger than the population size. I'm fine going ahead and computing the population value when the k=n. That is what the previous code did.

Updated to raise for out of bounds. I don't recall why I changed that recently; it must have been late and I was thoughtless. I added tests to guard against such thoughtlessness in the future.

dschult reviewed (approved):
This looks great!  Thanks @eriknw  and @ChuckHastings!

I approve this PR

rossbar reviewed (approved):
Thanks so much @eriknw @ChuckHastings and @dschult for the iterations and helpful discussion. This is indeed a nuanced issue! The updates LGTM and the computations were clearer than before, so let's get this in! As noted in the discussion, any additional followups (e.g. k==n corner cases etc.) can be handled in different PRs for a more focused discussion.

Thanks again!

dschult commented on networkx/algorithms/centrality/betweenness.py:
```diff
@@ -127,6 +127,12 @@ def betweenness_centrality(
        https://doi.org/10.2307/3033543
     """
     betweenness = dict.fromkeys(G, 0.0)  # b[v]=0 for v in G
+    if k == len(G):
+        # This is for performance and correctness. When `endpoints` is False
```
Just to close the loop on this comment/PR-discussion:

The previous behavior for `k<=n and endpoints=False` was incorrect. So we should not deprecate these changes. But we **should** add a paragraph to the release notes talking about it.  The problem was our scaling up from a sample of `k` nodes to the full BC values. This PR almost corrects everything. The next PR #7949 fixes the adjustment for whether a node is in the sample or not. That fix makes `k=n` work fine -- it could create a sample tha includes every single node and the result would be the same as `k=None`.  Even so, we still change `k==n` to `k=None` there because the resulting bc values are identical either way, and `k=None` is faster because we know we have to do all the nodes (so no setup of a sample or special scaling up to full results at the end).

:)
